From c84f5ca05307c5722be2ecc7f113e1f7be33c211 Mon Sep 17 00:00:00 2001
From: Edgar Chen <echen@mozilla.com>
Date: Tue, 25 Dec 2012 09:47:05 +0800
Subject: [PATCH] Bug 809725 - Part 3: Export Email to contact. r=gwagner

---
 dom/contacts/ContactManager.js | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/dom/contacts/ContactManager.js b/dom/contacts/ContactManager.js
index 9c49ec0..6b8993d 100644
--- a/dom/contacts/ContactManager.js
+++ b/dom/contacts/ContactManager.js
@@ -385,7 +385,13 @@ ContactManager.prototype = {
         if (req) {
           let result = contacts.map(function(c) {
             let contact = new Contact();
-            contact.init( { name: [c.alphaId], tel: [ { value: c.number } ] } );
+            let prop = {name: [c.alphaId], tel: [ { value: c.number } ]};
+
+            if (c.email) {
+              prop.email = [{value: c.email}];
+            }
+
+            contact.init(prop);
             return contact;
           });
           if (DEBUG) debug("result: " + JSON.stringify(result));
--
1.8.0

