diff --git a/cmds/app_process/Android.mk b/cmds/app_process/Android.mk
index 2391b72..6f033b5 100644
--- a/cmds/app_process/Android.mk
+++ b/cmds/app_process/Android.mk
@@ -1,5 +1,6 @@
 LOCAL_PATH:= $(call my-dir)
 include $(CLEAR_VARS)
+LOCAL_MODULE_TAGS := optional
 
 LOCAL_SRC_FILES:= \
 	app_main.cpp
diff --git a/cmds/surfaceflinger/Android.mk b/cmds/surfaceflinger/Android.mk
index 1df32bb..aa290b8 100644
--- a/cmds/surfaceflinger/Android.mk
+++ b/cmds/surfaceflinger/Android.mk
@@ -1,5 +1,6 @@
 LOCAL_PATH:= $(call my-dir)
 include $(CLEAR_VARS)
+LOCAL_MODULE_TAGS := optional
 
 LOCAL_SRC_FILES:= \
 	main_surfaceflinger.cpp 
diff --git a/include/utils/String16.h b/include/utils/String16.h
index 07a0c11..1a6ce23 100644
--- a/include/utils/String16.h
+++ b/include/utils/String16.h
@@ -26,8 +26,10 @@
 // ---------------------------------------------------------------------------
 
 extern "C" {
-
+// char32_t and char16_t are built-in types as of c++0x.
+#if !defined(__GXX_EXPERIMENTAL_CXX0X__) && __cplusplus < 201103L
 typedef uint16_t char16_t;
+#endif
 
 // Standard string functions on char16 strings.
 int strcmp16(const char16_t *, const char16_t *);
diff --git a/include/utils/String8.h b/include/utils/String8.h
index ef0b51a..8917e5c 100644
--- a/include/utils/String8.h
+++ b/include/utils/String8.h
@@ -30,8 +30,10 @@
 // ---------------------------------------------------------------------------
 
 extern "C" {
-
+// char32_t and char16_t are built-in types as of c++0x.
+#if !defined(__GXX_EXPERIMENTAL_CXX0X__) && __cplusplus < 201103L
 typedef uint32_t char32_t;
+#endif
 
 size_t strlen32(const char32_t *);
 size_t strnlen32(const char32_t *, size_t);
-- 
1.7.8.3

