diff --git a/patch_downloader/Android.mk b/patch_downloader/Android.mk
index 89df2e1..ba58ed8 100755
--- a/patch_downloader/Android.mk
+++ b/patch_downloader/Android.mk
@@ -18,7 +18,7 @@ LOCAL_C_INCLUDES :=  \
 
 LOCAL_SHARED_LIBRARIES := libqmi libqcci_legacy libqmiservices libcutils
 LOCAL_SRC_FILES := fm_qsoc_patches.c
-LOCAL_MODULE_TAGS := optional eng
+LOCAL_MODULE_TAGS := optional
 
 LOCAL_ADDITIONAL_DEPENDENCIES := $(TARGET_OUT_INTERMEDIATES)/KERNEL_OBJ/usr
 
