diff --git a/core/config.mk b/core/config.mk
index 41e60ea..5797392 100644
--- a/core/config.mk
+++ b/core/config.mk
@@ -100,7 +100,7 @@ COMMON_JAVA_PACKAGE_SUFFIX := .jar
 COMMON_ANDROID_PACKAGE_SUFFIX := .apk
 
 # list of flags to turn specific warnings in to errors
-TARGET_ERROR_FLAGS := -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point
+TARGET_ERROR_FLAGS := -Werror=return-type -Werror=address -Werror=sequence-point
 
 # TODO: do symbol compression
 TARGET_COMPRESS_MODULE_SYMBOLS := false
