diff --git a/common.mk b/common.mk
index e616489..3a3b1cd 100644
--- a/common.mk
+++ b/common.mk
@@ -170,3 +170,7 @@ PRODUCT_PROPERTY_OVERRIDES += \
     ro.vendor.extension_library=/system/lib/libqc-opt.so
 
 PRODUCT_PRIVATE_KEY := device/qcom/common/qcom.key
+
+$(call inherit-product, device/qcom/b2g_common/b2g_product.mk)
+include device/qcom/b2g_common/keymaps/keymaps.mk
+TARGET_PROVIDES_B2G_INIT_RC=true
-- 
1.7.8.3

