diff --git a/core/java.mk b/core/java.mk
index 1cde62b..41d4649 100644
--- a/core/java.mk
+++ b/core/java.mk
@@ -194,6 +194,7 @@ endif
 # common_java.mk, and moving the include of base_rules.mk to
 # after all the declarations.
 
+LOCAL_UNINSTALLABLE_MODULE:=true
 #######################################
 include $(BUILD_SYSTEM)/base_rules.mk
 #######################################
diff --git a/core/notice_files.mk b/core/notice_files.mk
index 74beb49..06621d4 100644
--- a/core/notice_files.mk
+++ b/core/notice_files.mk
@@ -34,7 +34,7 @@ else
       module_installed_filename := \
           $(patsubst $(PRODUCT_OUT)%,%,$($(my_prefix)OUT_JAVA_LIBRARIES))/$(module_leaf)
     else
-      $(error Cannot determine where to install NOTICE file for $(LOCAL_MODULE))
+      $(info Cannot determine where to install NOTICE file for $(LOCAL_MODULE))
     endif # JAVA_LIBRARIES
   endif # STATIC_LIBRARIES
 endif
diff --git a/target/product/core.mk b/target/product/core.mk
index abb2d9f..c3c6e9e 100644
--- a/target/product/core.mk
+++ b/target/product/core.mk
@@ -23,46 +23,11 @@ PRODUCT_PROPERTY_OVERRIDES := \
     ro.config.alarm_alert=Alarm_Classic.ogg
 
 PRODUCT_PACKAGES := \
-    ApplicationsProvider \
-    BackupRestoreConfirmation \
-    Browser \
-    Contacts \
-    ContactsProvider \
-    DefaultContainerService \
-    DownloadProvider \
-    DownloadProviderUi \
-    HTMLViewer \
-    Home \
-    KeyChain \
-    MediaProvider \
-    PackageInstaller \
-    PicoTts \
-    SettingsProvider \
-    SharedStorageBackup \
-    TelephonyProvider \
-    UserDictionaryProvider \
-    VpnDialogs \
-    apache-xml \
-    bouncycastle \
     bu \
     cacerts \
-    com.android.location.provider \
-    com.android.location.provider.xml \
-    core \
-    core-junit \
-    dalvikvm \
-    dexdeps \
-    dexdump \
-    dexlist \
-    dexopt \
     dmtracedump \
     dx \
-    ext \
-    filterfw \
-    framework-res \
     hprof-conv \
-    icu.dat \
-    installd \
     ip \
     ip-up-vpn \
     ip6tables \
diff --git a/target/product/full_base.mk b/target/product/full_base.mk
index d89af51..e220778 100644
--- a/target/product/full_base.mk
+++ b/target/product/full_base.mk
@@ -20,9 +20,6 @@
 # in inherited configurations.
 
 PRODUCT_PACKAGES := \
-    drmserver \
-    libdrmframework \
-    libdrmframework_jni \
     libfwdlockengine \
     OpenWnn \
     PinyinIME \
diff --git a/target/product/generic_no_telephony.mk b/target/product/generic_no_telephony.mk
index 5169f4e..4a4d57e 100644
--- a/target/product/generic_no_telephony.mk
+++ b/target/product/generic_no_telephony.mk
@@ -48,9 +48,6 @@ PRODUCT_PACKAGES := \
     wpa_supplicant.conf
 
 PRODUCT_PACKAGES += \
-    icu.dat
-
-PRODUCT_PACKAGES += \
     librs_jni \
     libvideoeditor_jni \
     libvideoeditorplayer \
-- 
1.7.8.3

