diff --git a/rootdir/etc/init.qcom.usb.sh b/rootdir/etc/init.qcom.usb.sh
index 82d8eca..3c31cd7 100644
--- a/rootdir/etc/init.qcom.usb.sh
+++ b/rootdir/etc/init.qcom.usb.sh
@@ -104,7 +104,11 @@ case "$usb_config" in
                 esac
             ;;
             "msm7627a")
-                setprop persist.sys.usb.config diag,serial_smd,serial_tty,rmnet_smd,mass_storage,adb
+                if [ -z $usb_config ]; then
+                     setprop persist.sys.usb.config diag,serial_smd,serial_tty,rmnet_smd,mass_storage
+                else
+                     setprop persist.sys.usb.config diag,serial_smd,serial_tty,rmnet_smd,mass_storage,adb
+                fi
             ;;
             * )
                 case "$baseband" in
