diff --git a/AndroidBoot.mk b/AndroidBoot.mk
index 8854579..ef45c32 100644
--- a/AndroidBoot.mk
+++ b/AndroidBoot.mk
@@ -3,7 +3,7 @@
 ifeq ($(call is-android-codename,HONEYCOMB),true)
 CROSS_TOOL := ../../../prebuilt/linux-x86/toolchain/arm-eabi-4.4.3/bin/arm-eabi-
 else ifeq ($(call is-android-codename,ICECREAM_SANDWICH),true)
-CROSS_TOOL := ../../../prebuilt/linux-x86/toolchain/arm-eabi-4.4.3/bin/arm-eabi-
+CROSS_TOOL := $(ANDROID_TOOLCHAIN)/arm-linux-androideabi-
 else
 CROSS_TOOL := ../../../prebuilt/linux-x86/toolchain/arm-eabi-4.4.0/bin/arm-eabi-
 endif
