diff --git a/apps/video/js/video.js b/apps/video/js/video.js
index 8da0099..2ce6a23 100644
--- a/apps/video/js/video.js
+++ b/apps/video/js/video.js
@@ -638,8 +638,11 @@ function setPlayerSize() {
   width *= scale;
   height *= scale;
 
-  var left = ((containerWidth - width) / 2);
-  var top = ((containerHeight - height) / 2);
+  var left = ((containerWidth - width).toFixed(9) / 2);
+  var top = ((containerHeight - height).toFixed(9) / 2);
+
+  // To remove the useless pending 0 of toFixed API call.
+  scale = 0 + scale.toFixed(9);
 
   var transform;
   switch (rotation) {
@@ -648,17 +651,18 @@ function setPlayerSize() {
     break;
   case 90:
     transform =
-      'translate(' + (left + width) + 'px,' + top + 'px) ' +
+      'translate(' + (left + width.toFixed(9)) + 'px,' + top + 'px) ' +
       'rotate(90deg)';
     break;
   case 180:
     transform =
-      'translate(' + (left + width) + 'px,' + (top + height) + 'px) ' +
+      'translate(' + (left + width.toFixed(9)) + 'px,' +
+                     (top + height.toFixed(9)) + 'px) ' +
       'rotate(180deg)';
     break;
   case 270:
     transform =
-      'translate(' + left + 'px,' + (top + height) + 'px) ' +
+      'translate(' + left + 'px,' + (top + height.toFixed(9)) + 'px) ' +
       'rotate(270deg)';
     break;
   }
