diff --git a/flash.sh b/flash.sh
index fe7d3cd..d35efcb 100755
--- a/flash.sh
+++ b/flash.sh
@@ -90,6 +90,11 @@ case "$1" in
 	make -C gaia install-gaia
 	exit $?
 	;;
+
+"time")
+	update_time
+	exit $?
+	;;
 esac
 
 case "$DEVICE" in
@@ -108,7 +113,11 @@ case "$DEVICE" in
 	;;
 
 *)
-	echo Unsupported device \"$DEVICE\", can\'t flash
-	exit -1
+	if [[ $(type -t flash_${DEVICE}) = function ]]; then
+		flash_${DEVICE} $1
+	else
+		echo Unsupported device \"$DEVICE\", can\'t flash
+		exit -1
+	fi
 	;;
 esac
diff --git a/run-gdb.sh b/run-gdb.sh
index 3a44fbd..4ade5c4 100755
--- a/run-gdb.sh
+++ b/run-gdb.sh
@@ -8,7 +8,7 @@ fi
 
 ADB=adb
 GDB_PORT=22576
-GDB=prebuilt/`uname -s | tr "[[:upper:]]" "[[:lower:]]"`-x86/toolchain/arm-linux-androideabi-4.4.x/bin/arm-linux-androideabi-gdb
+: ${GDB:=prebuilt/`uname -s | tr "[[:upper:]]" "[[:lower:]]"`-x86/toolchain/arm-linux-androideabi-4.4.x/bin/arm-linux-androideabi-gdb}
 B2G_BIN=/system/b2g/b2g
 GDBINIT=/tmp/b2g.gdbinit.`whoami`
 
-- 
1.7.8.3

