diff --git a/Android.mk b/Android.mk
index 82ca524..55f2449 100644
--- a/Android.mk
+++ b/Android.mk
@@ -25,6 +25,13 @@ GAIA_MAKE_FLAGS += PRODUCTION=1
 B2G_SYSTEM_APPS := 1
 endif
 
+# Disable the First time user experience in engineering builds
+ifneq ($(filter eng, $(TARGET_BUILD_VARIANT)),)
+ifeq ($(FTU),)
+GAIA_MAKE_FLAGS += NOFTU=1
+endif
+endif
+
 ifeq ($(B2G_SYSTEM_APPS),1)
 GAIA_MAKE_FLAGS += B2G_SYSTEM_APPS=1
 GAIA_APP_INSTALL_PARENT := $(TARGET_OUT)/b2g
