diff --git a/Makefile b/Makefile
index 0a8cfe6..5ed8caa 100644
--- a/Makefile
+++ b/Makefile
@@ -32,7 +32,7 @@ DEBUG?=0
 
 REPORTER=Spec
 
-GAIA_APP_SRCDIRS?=apps test_apps showcase_apps
+GAIA_APP_SRCDIRS?=$(CURDIR)/apps $(CURDIR)/test_apps $(CURDIR)/showcase_apps
 
 GAIA_ALL_APP_SRCDIRS=$(GAIA_APP_SRCDIRS)
 
@@ -40,9 +40,9 @@ GAIA_APP_RELATIVEPATH=$(foreach dir, $(GAIA_APP_SRCDIRS), $(wildcard $(dir)/*))
 
 ifeq ($(MAKECMDGOALS), demo)
 GAIA_DOMAIN=thisdomaindoesnotexist.org
-GAIA_APP_SRCDIRS=apps showcase_apps
+GAIA_APP_SRCDIRS=$(CURDIR)/apps $(CURDIR)/showcase_apps
 else ifeq ($(MAKECMDGOALS), production)
-GAIA_APP_SRCDIRS=apps
+GAIA_APP_SRCDIRS=$(CURDIR)/apps
 endif
 
 
diff --git a/build/offline-cache.js b/build/offline-cache.js
index 39a7e14..292412e 100644
--- a/build/offline-cache.js
+++ b/build/offline-cache.js
@@ -71,8 +71,7 @@ function storeCache(id, url, content, count) {
 function getSubDirectories(directory) {
   let appsDir = Cc["@mozilla.org/file/local;1"]
                .createInstance(Ci.nsILocalFile);
-  appsDir.initWithPath(GAIA_DIR);  
-  appsDir.append(directory);
+  appsDir.initWithPath(directory);
 
   let dirs = [];  
   let files = appsDir.directoryEntries;  
@@ -158,8 +157,7 @@ function getFileContent(file) {
 function getJSON(root, dir, name) {
   let file = Cc["@mozilla.org/file/local;1"]
                .createInstance(Ci.nsILocalFile);
-  file.initWithPath(GAIA_DIR);
-  file.append(root);
+  file.initWithPath(root);
   file.append(dir);
   file.append(name);
 
@@ -201,8 +199,7 @@ appSrcDirs.forEach(function parseDirectory(directoryName) {
     let domain = "http://" + dir + "." + GAIA_DOMAIN;
     let manifest = Cc["@mozilla.org/file/local;1"]
                    .createInstance(Ci.nsILocalFile);
-    manifest.initWithPath(GAIA_DIR);
-    manifest.append(directoryName);
+    manifest.initWithPath(directoryName);
 
     const base = manifest.path;
     manifest.append(dir);
diff --git a/build/preferences.js b/build/preferences.js
index 6c1c85f..04df482 100644
--- a/build/preferences.js
+++ b/build/preferences.js
@@ -4,8 +4,7 @@ const { 'classes': Cc, 'interfaces': Ci, 'results': Cr, } = Components;
 function getSubDirectories(directory) {
   let appsDir = Cc["@mozilla.org/file/local;1"]
                .createInstance(Ci.nsILocalFile);
-  appsDir.initWithPath(GAIA_DIR);
-  appsDir.append(directory);
+  appsDir.initWithPath(directory);
 
   let dirs = [];
   let files = appsDir.directoryEntries;
@@ -42,8 +41,7 @@ function getFileContent(file) {
 function getJSON(root, dir, name) {
   let file = Cc["@mozilla.org/file/local;1"]
                .createInstance(Ci.nsILocalFile);
-  file.initWithPath(GAIA_DIR);
-  file.append(root);
+  file.initWithPath(root);
   file.append(dir);
   file.append(name);
 
