diff --git a/Makefile b/Makefile
index b113d1d..565ebbb 100644
--- a/Makefile
+++ b/Makefile
@@ -175,7 +175,7 @@ webapp-manifests:
 
 
 # Generate profile/OfflineCache/
-offline: install-xulrunner
+offline: install-xulrunner update-offline-manifests webapp-manifests
 ifneq ($(DEBUG),1)
 	@echo "Building offline cache"
 	@rm -rf profile/OfflineCache
@@ -432,7 +432,7 @@ forward:
 
 
 # update the manifest.appcache files to match what's actually there
-update-offline-manifests:
+update-offline-manifests: stamp-commit-hash
 	for d in `find ${GAIA_APP_SRCDIRS} -mindepth 1 -maxdepth 1 -type d` ;\
 	do \
 		rm -rf $$d/manifest.appcache ;\
-- 
1.7.8.3

