diff --git a/include/media/stagefright/OMXCodec.h b/include/media/stagefright/OMXCodec.h
index d1444b4..ad563ee 100755
--- a/include/media/stagefright/OMXCodec.h
+++ b/include/media/stagefright/OMXCodec.h
@@ -209,10 +209,12 @@ private:
     bool mOutputPortSettingsChangedPending;
     bool mThumbnailMode;
 
+    uint32_t mNumClientBuffers;
     MediaBuffer *mLeftOverBuffer;
 
     Mutex mLock;
     Condition mAsyncCompletion;
+    Condition mBufferCompletion;
 
     bool mPaused;
 
diff --git a/media/libstagefright/OMXCodec.cpp b/media/libstagefright/OMXCodec.cpp
index 7a57834..1ff515e 100644
--- a/media/libstagefright/OMXCodec.cpp
+++ b/media/libstagefright/OMXCodec.cpp
@@ -2143,6 +2143,7 @@ OMXCodec::OMXCodec(
       mSeekMode(ReadOptions::SEEK_CLOSEST_SYNC),
       mTargetTimeUs(-1),
       mOutputPortSettingsChangedPending(false),
+      mNumClientBuffers(0),
       mLeftOverBuffer(NULL),
       mPaused(false),
       mIsAacFormatAdif(0),
@@ -4909,6 +4910,10 @@ status_t OMXCodec::stop() {
         mAsyncCompletion.wait(mLock);
     }
 
+    while (mNumClientBuffers) {
+        mBufferCompletion.wait(mLock);
+    }
+
     bool isError = false;
     switch (mState) {
         case LOADED:
@@ -5215,6 +5220,7 @@ status_t OMXCodec::read(
 
     BufferInfo *info = &mPortBuffers[kPortIndexOutput].editItemAt(index);
     CHECK_EQ((int)info->mStatus, (int)OWNED_BY_US);
+    mNumClientBuffers++;
     info->mStatus = OWNED_BY_CLIENT;
 
     info->mMediaBuffer->add_ref();
@@ -5234,7 +5240,9 @@ void OMXCodec::signalBufferReturned(MediaBuffer *buffer) {
             CHECK_EQ((int)mPortStatus[kPortIndexOutput], (int)ENABLED);
             CHECK_EQ((int)info->mStatus, (int)OWNED_BY_CLIENT);
 
+            mNumClientBuffers--;
             info->mStatus = OWNED_BY_US;
+            mBufferCompletion.signal();
 
             if (buffer->graphicBuffer() == 0) {
                 fillOutputBuffer(info);
